### DISCUSSION TITLE
Input from code: I am struggling

### DISCUSSION THREAD

**Discussion by LeannAlexandra** [libretro/docs#875] [Q&A]
<h1>How can I simulate user-input from another project programatically?</h1>
firstly: if this is the wrong platform- please direct me to the right person to talk to. 
<h4>Background</h4>
Hi, I wrote some base code for an ai to play games using RetroArch. So far I have a foundation program in c++ using OpenCV  to extract the matrix, Tesseract for scores & timers (to rate the progress)  and planning on  tensorflow to be the Neural Network to process the frames: Here is the current state of progress

![Screenshot_20230914_215402](https://github.com/libretro/docs/assets/81232303/0a4e27e6-ee62-44dd-875c-41f896a543a3)

<h4>What I have tried</h4>
My current solution bogs down my system to simulate global input - but it can not run in the background, and I cannot do anything else. So here are my current considerations: 

<h4>Current Options</h4>
<ol><li>
Make an abomination of my fork of RetroArch to make it unplayable by human. - (not inline with goals)
<ol><li>This might even help my interpretation for assisted AI by using the information needed before drawing the background, which takes away from the "ooooooh loook my shiny smart ai" hype</li>
<li>This can also help in quickly iterating to train ai's</li>
</ol>
</li>
<li>simulate global input, run in a virtual machine or docker image - feels like unnecessary layer of virtualization. - will use more resources than previous method.</li>

<li>ask for help from forums to achieve the first point for training but still stay true to the original idea <em>(fully knowing I'd be looking like an idiot)*</em> </li>

</ol>

<h4>Current Challenges</h4>
<ul>
<li>My knowledge of C is limited</li>
 <li>I need to understand how to use a possible API-call to simulate input, even if the window is in the background.</li> 
</ul>

<h4>Desired Outcome</h4>
<ul>
<li>Create General AI that can play games based on visual input - because studying programming leaves me without time to play the games, therefore I need an AI that can play games in my stead. </li>
<li>Seriously I would like to know how to turn the nobs under the hood - to one day create better ai and npc's in the future - but till then, I just want to learn and contribute </li>
</ul>

How to reach me? - [github](https://github.com/leannalexandra), [email](mailto:leannalexandraviolet@gmail.com)

**Comment by hizzlekizzle** [comment]
RetroArch has a command interface where you can send packets via UDP socket to make stuff happen (load games, save state, etc). I don't think inputs are included currently, but that may be an adequate path for your purposes.

**Reply by LeannAlexandra** [reply]
I checked it out, not exactly what I needed. - I think I'll need a completely virtual joystick

**Reply by LeannAlexandra** [reply]
update: I did write a virtual controller that is registered globally as if a physical device is plugged in that can be controlled by the AI thanks for the assistance

**Comment by LeannAlexandra** [comment]
CLOSING REMARK: Wrote a global virtual Controller to do as intended.

**Comment by hizzlekizzle** [comment]
Very cool. Is that published publicly anywhere?

**Reply by LeannAlexandra** [reply]
### Yes, Yes it does
there is a general version of it here: [virtualController](https://github.com/LeannAlexandra/virtualController/blob/main/src/virtualController.cpp) - just a few dents nicks and scratches need to be polished out - after this repository proved my concept I moved on to the project using it. [edit:  **aiplayercontroller.hpp** is the best place to start]

### 'works on my system' - Manjaro

1. Part 1: main.c duplicates an existing controller (not dynamically sourced yet, but I could, and did use libudev to source it previously), the other emulator (mGBA) would snatch my device off the eventX list (basically leaving the device there but without the "eventX handle"  then in that case, libudev & EVIOCGRAB would not suffice in stopping this from happening. - It works fine with libretro (on my system). with a little more elbow grease I could potentially read a second controller, and program the xbox controller button to switch modes nand, or, xor and send those out via my upgraded concept -> part two:   
2. Part 2: (the solution) 
This is now a library class, which can be used to create a controller without input device cloning -
src/virtualcontroller.cpp (and header include/virtualcontroller.hpp) registers a global device (literally a virtual clone of an xbox controller even in vendor and product id )- but it registers as "AI Player Controller" which can be seen in linux inputs too and now exist as a systemwide (virtual) plugged in device. 

To be honest, there is much we can 'add and refine' from it - and this version is terrible - it doesn't really exist as an api yet, because a neural network doesn't really care about button 'labels' so it exists as a struct of 8 abs values and 11 buttons - I am pretty sure a thing like a switch would have some more buttons, and motion capture too, so from the basis of this class, you'd be close enough to make it work. 

### The Bigger Picture
I am actually more exited about the bigger project that is using this class than  I am actually expanding this class, as I feel in the bigger picture the sub part becomes novel*

The idea was to jump the hype-train of ai, and make one that can eventually learn to copilot something like ice climber. Think of it as your little sibling watching you play, <img src="https://i.pinimg.com/originals/fb/74/63/fb7463a70b021c87239bb60a27e9d1a5.png"> you give them this controller (and that is their world) and when you step away this little ai steps up and completely sucks (at first) but the more they watch you play the better they become. 
Not a bad concept ? - I just cannot get past my (core dumped) obsession it seems. (my focus now is pytorch (c++) the final 'piece' of the puzzle, then I'll make it public too:   [aiplayer](https://github.com/LeannAlexandra/aiplayer)

**Reply by hizzlekizzle** [reply]
That sounds awesome man :D
